# Data Science Graded Challenge 1 Phase 1

by: Muhammad Farhan Zahirsyah

Batch 10

---
### Desctription
Analisis regresi mempelajari bentuk hubungan antara satu atau lebih pengubah/variabel bebas (X) dengan satu pengubah tak bebas (Y). Dalam penelitian pengubah bebas (X) biasanya pengubah yang ditentukan oleh peneliti secara bebas misalnya dosis obat, lama penyimpanan, kadar zat pengawet, umur ternak dan sebagainya. Disamping itu pengubah bebas bisa juga berupa pengubah tak bebasnya, misalnya dalam pengukuran panjang badan dan berat badan sapi, karena panjang badan lebih mudah diukur maka panjang badan dimasukkan kedalam pengubah bebas (X), sedangkan berat badan dimasukkan pengubah tak bebas (Y). (Hidayat, 2012)

Uber and Lyft Dataset Boston, Massachusetts berisi mengenai semua telemetri dari suatu order baik dari aplikasi lyft maupun dari aplikasi Uber. dataset ini juga lengkap memberikan informasi tentang kondisi cuaca saat itu. Data ini bisa mempengaruhi biaya jasa menggunakan taksi online Uber atau Lyft. salah satu cara memprediksi biaya taksi online adalah menggunakan Regresi yang akan dilakukan pada penelitian ini.


### Objective
- Memprediksi biaya jasa taksi online Uber dan Lyft


### Problem Statement
- bagaimana cara memprediksi biaya taksi online Uber dan Lyft


---


## Import Library

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import matplotlib
import statsmodels.api as sm
import sklearn
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

In [2]:
print('Pandas Version     :',pd.__version__)
print('Numpy Version      :',np.__version__)
print('Statsmodels Version:',sm.__version__)
print('Matplotlib Version :',matplotlib.__version__)
print('Seaborn Version    :',sns.__version__)
print('Sklearn Version    :',sklearn.__version__)

Pandas Version     : 1.3.4
Numpy Version      : 1.20.3
Statsmodels Version: 0.12.2
Matplotlib Version : 3.4.3
Seaborn Version    : 0.11.2
Sklearn Version    : 0.24.2


## Data Loading

In [3]:
uber_lyft_df = pd.read_csv('../big_dataset/rideshare_kaggle.csv')

In [10]:
uber_lyft_df.head()

id     timestamp  hour  day  month  \
0  424553bb-7174-41ea-aeb4-fe06d4f4b9d7  1.544953e+09     9   16     12   
1  4bd23055-6827-41c6-b23b-3c491f24e74d  1.543284e+09     2   27     11   
2  981a3613-77af-4620-a42a-0c0866077d1e  1.543367e+09     1   28     11   
3  c2d88af2-d278-4bfd-a8d0-29ca77cc5512  1.543554e+09     4   30     11   
4  e0126e1f-8ca9-4f2e-82b3-50505a09db9a  1.543463e+09     3   29     11   

              datetime          timezone            source    destination  \
0  2018-12-16 09:30:07  America/New_York  Haymarket Square  North Station   
1  2018-11-27 02:00:23  America/New_York  Haymarket Square  North Station   
2  2018-11-28 01:00:22  America/New_York  Haymarket Square  North Station   
3  2018-11-30 04:53:02  America/New_York  Haymarket Square  North Station   
4  2018-11-29 03:49:20  America/New_York  Haymarket Square  North Station   

  cab_type  ... precipIntensityMax uvIndexTime  temperatureMin  \
0     Lyft  ...             0.1276  1544979600           39.89   
1     Lyft  ...             0.1300  1543251600           40.49   
2     Lyft  ...             0.1064  1543338000           35.36   
3     Lyft  ...             0.0000  1543507200           34.67   
4     Lyft  ...             0.0001  1543420800           33.10   

   temperatureMinTime  temperatureMax  temperatureMaxTime  \
0          1545012000           43.68          1544968800   
1          1543233600           47.30          1543251600   
2          1543377600           47.55          1543320000   
3          1543550400           45.03          1543510800   
4          1543402800           42.18          1543420800   

   apparentTemperatureMin  apparentTemperatureMinTime  apparentTemperatureMax  \
0                   33.73                  1545012000                   38.07   
1                   36.20                  1543291200                   43.92   
2                   31.04                  1543377600                   44.12   
3                   30.30                  1543550400                   38.53   
4                   29.11                  1543392000                   35.75   

  apparentTemperatureMaxTime  
0                 1544958000  
1                 1543251600  
2                 1543320000  
3                 1543510800  
4                 1543420800  

[5 rows x 57 columns]

diatas adalah 5 baris pertama dataset

In [11]:
uber_lyft_df.head(3).T

0  \
id                           424553bb-7174-41ea-aeb4-fe06d4f4b9d7   
timestamp                                           1544952607.89   
hour                                                            9   
day                                                            16   
month                                                          12   
datetime                                      2018-12-16 09:30:07   
timezone                                         America/New_York   
source                                           Haymarket Square   
destination                                         North Station   
cab_type                                                     Lyft   
product_id                                              lyft_line   
name                                                       Shared   
price                                                         5.0   
distance                                                     0.44   
surge_multiplier                                              1.0   
latitude                                                  42.2148   
longitude                                                 -71.033   
temperature                                                 42.34   
apparentTemperature                                         37.12   
short_summary                                      Mostly Cloudy    
long_summary                            Rain throughout the day.    
precipIntensity                                               0.0   
precipProbability                                             0.0   
humidity                                                     0.68   
windSpeed                                                    8.66   
windGust                                                     9.17   
windGustTime                                           1545015600   
visibility                                                   10.0   
temperatureHigh                                             43.68   
temperatureHighTime                                    1544968800   
temperatureLow                                              34.19   
temperatureLowTime                                     1545048000   
apparentTemperatureHigh                                     37.95   
apparentTemperatureHighTime                            1544968800   
apparentTemperatureLow                                      27.39   
apparentTemperatureLowTime                             1545044400   
icon                                         partly-cloudy-night    
dewPoint                                                     32.7   
pressure                                                  1021.98   
windBearing                                                    57   
cloudCover                                                   0.72   
uvIndex                                                         0   
visibility.1                                                 10.0   
ozone                                                       303.8   
sunriseTime                                            1544962084   
sunsetTime                                             1544994864   
moonPhase                                                     0.3   
precipIntensityMax                                         0.1276   
uvIndexTime                                            1544979600   
temperatureMin                                              39.89   
temperatureMinTime                                     1545012000   
temperatureMax                                              43.68   
temperatureMaxTime                                     1544968800   
apparentTemperatureMin                                      33.73   
apparentTemperatureMinTime                             1545012000   
apparentTemperatureMax                                      38.07   
apparentTemperatureMaxTime                             1544958000   

                                                             

Diatas menampilkan 3 data teratas dengan menampilkan nama seluruh kolom

In [5]:
uber_lyft_df.shape

(693071, 57)

Terdapat 693071 baris data dengan 57 kolom

In [53]:
# mengelompokan data berdasarkan datatype
num_columns = uber_lyft_df.select_dtypes(include=np.number).columns.tolist()
obj_columns = uber_lyft_df.select_dtypes(include=['object']).columns.tolist()
uber_lyft_df[num_columns].describe().T

count          mean            std  \
timestamp                    693071.0  1.544046e+09  689192.492586   
hour                         693071.0  1.161914e+01       6.948114   
day                          693071.0  1.779436e+01       9.982286   
month                        693071.0  1.158668e+01       0.492429   
price                        637976.0  1.654513e+01       9.324359   
distance                     693071.0  2.189430e+00       1.138937   
surge_multiplier             693071.0  1.013870e+00       0.091641   
latitude                     693071.0  4.233817e+01       0.047840   
longitude                    693071.0 -7.106615e+01       0.020302   
temperature                  693071.0  3.958439e+01       6.726084   
apparentTemperature          693071.0  3.588122e+01       7.918707   
precipIntensity              693071.0  8.922153e-03       0.026901   
precipProbability            693071.0  1.460758e-01       0.328938   
humidity                     693071.0  7.411187e-01       0.138595   
windSpeed                    693071.0  6.186253e+00       3.148223   
windGust                     693071.0  8.469860e+00       5.289179   
windGustTime                 693071.0  1.544049e+09  692824.402314   
visibility                   693071.0  8.467970e+00       2.602888   
temperatureHigh              693071.0  4.504098e+01       5.996541   
temperatureHighTime          693071.0  1.544050e+09  693792.107469   
temperatureLow               693071.0  3.415074e+01       6.383163   
temperatureLowTime           693071.0  1.544102e+09  692292.331983   
apparentTemperatureHigh      693071.0  4.161405e+01       7.666138   
apparentTemperatureHighTime  693071.0  1.544050e+09  694169.873816   
apparentTemperatureLow       693071.0  3.013982e+01       8.057468   
apparentTemperatureLowTime   693071.0  1.544099e+09  692737.825154   
dewPoint                     693071.0  3.165626e+01       9.142355   
pressure                     693071.0  1.010095e+03      13.472899   
windBearing                  693071.0  2.200559e+02      99.102736   
cloudCover                   693071.0  6.865020e-01       0.358534   
uvIndex                      693071.0  2.489081e-01       0.474097   
visibility.1                 693071.0  8.467970e+00       2.602888   
ozone                        693071.0  3.135136e+02      27.953061   
sunriseTime                  693071.0  1.544027e+09  691139.271721   
sunsetTime                   693071.0  1.544060e+09  690663.392963   
moonPhase                    693071.0  5.790655e-01       0.244705   
precipIntensityMax           693071.0  3.737441e-02       0.055214   
uvIndexTime                  693071.0  1.544044e+09  691202.767271   
temperatureMin               693071.0  3.345777e+01       6.467224   
temperatureMinTime           693071.0  1.544042e+09  690195.441406   
temperatureMax               693071.0  4.526131e+01       5.645046   
temperatureMaxTime           693071.0  1.544047e+09  690135.337190   
apparentTemperatureMin       693071.0  2.973100e+01       7.110494   
apparentTemperatureMinTime   693071.0  1.544048e+09  687186.191495   
apparentTemperatureMax       693071.0  4.199734e+01       6.936841   
apparentTemperatureMaxTime   693071.0  1.544048e+09  691077.651976   

                                      min           25%           50%  \
timestamp                    1.543204e+09  1.543444e+09  1.543737e+09   
hour                         0.000000e+00  6.000000e+00  1.200000e+01   
day                          1.000000e+00  1.300000e+01  1.700000e+01   
month                        1.100000e+01  1.100000e+01  1.200000e+01   
price                        2.500000e+00  9.000000e+00  1.350000e+01   
distance                     2.000000e-02  1.280000e+00  2.160000e+00   
surge_multiplier             1.000000e+00  1.000000e+00  1.000000e+00   
latitude                     4.221480e+01  4.235030e+01  4.235190e+01   
longitude                   -7.110540e+01 -7.108100e+01 -7.106310e+01   
temperature

In [52]:
len(num_columns)

46

Terdapat 46 kolom berjenis number

In [7]:
uber_lyft_df[obj_columns].describe().T

count  unique                                   top    freq
id             693071  693071  424553bb-7174-41ea-aeb4-fe06d4f4b9d7       1
datetime       693071   31350                   2018-11-26 23:21:14     156
timezone       693071       1                      America/New_York  693071
source         693071      12                    Financial District   58857
destination    693071      12                    Financial District   58851
cab_type       693071       2                                  Uber  385663
product_id     693071      13  6f72dfc5-27f1-42e8-84db-ccc7a75f6969   55096
name           693071      13                                UberXL   55096
short_summary  693071       9                             Overcast   218895
long_summary   693071      11    Mostly cloudy throughout the day.   202340
icon           693071       7                               cloudy   218895

In [54]:
len(obj_columns)

11

terdapat 11 kolom dengan data types categorical 

In [8]:
for i in obj_columns[2:]:
    print(uber_lyft_df[i].unique())
    print("-"*50)

['America/New_York']
--------------------------------------------------
['Haymarket Square' 'Back Bay' 'North End' 'North Station' 'Beacon Hill'
 'Boston University' 'Fenway' 'South Station' 'Theatre District'
 'West End' 'Financial District' 'Northeastern University']
--------------------------------------------------
['North Station' 'Northeastern University' 'West End' 'Haymarket Square'
 'South Station' 'Fenway' 'Theatre District' 'Beacon Hill' 'Back Bay'
 'North End' 'Financial District' 'Boston University']
--------------------------------------------------
['Lyft' 'Uber']
--------------------------------------------------
['lyft_line' 'lyft_premier' 'lyft' 'lyft_luxsuv' 'lyft_plus' 'lyft_lux'
 '6f72dfc5-27f1-42e8-84db-ccc7a75f6969'
 '6c84fd89-3f11-4782-9b50-97c468b19529'
 '55c66225-fbe7-4fd5-9072-eab1ece5e23e'
 '9a0e7b09-b92b-4c41-9779-2ad22b4d779d'
 '6d318bcc-22a3-4af6-bddd-b409bfce1546'
 '997acbb5-e102-41e1-b155-9df7de0a73f2'
 '8cf7e821-f0d3-49c6-8eba-e679c0ebcf6a']
----------

Diatas adalah isi dari kolom yang berdatatypes categorical

In [9]:
uber_lyft_df.isnull().sum()

id                                 0
timestamp                          0
hour                               0
day                                0
month                              0
datetime                           0
timezone                           0
source                             0
destination                        0
cab_type                           0
product_id                         0
name                               0
price                          55095
distance                           0
surge_multiplier                   0
latitude                           0
longitude                          0
temperature                        0
apparentTemperature                0
short_summary                      0
long_summary                       0
precipIntensity                    0
precipProbability                  0
humidity                           0
windSpeed                          0
windGust                           0
windGustTime                       0
v

terdapat missing 55095 di kolom price. ini bisa digunakan sebagai data test. namun dengan data ini tidak bisa dicek modelnya apabila digunakan data test tersebut. karena price nya hilang dan tidak bisa diketahui akurasi dari model tersebut.

In [12]:
uber_lyft_df.columns

Index(['id', 'timestamp', 'hour', 'day', 'month', 'datetime', 'timezone',
       'source', 'destination', 'cab_type', 'product_id', 'name', 'price',
       'distance', 'surge_multiplier', 'latitude', 'longitude', 'temperature',
       'apparentTemperature', 'short_summary', 'long_summary',
       'precipIntensity', 'precipProbability', 'humidity', 'windSpeed',
       'windGust', 'windGustTime', 'visibility', 'temperatureHigh',
       'temperatureHighTime', 'temperatureLow', 'temperatureLowTime',
       'apparentTemperatureHigh', 'apparentTemperatureHighTime',
       'apparentTemperatureLow', 'apparentTemperatureLowTime', 'icon',
       'dewPoint', 'pressure', 'windBearing', 'cloudCover', 'uvIndex',
       'visibility.1', 'ozone', 'sunriseTime', 'sunsetTime', 'moonPhase',
       'precipIntensityMax', 'uvIndexTime', 'temperatureMin',
       'temperatureMinTime', 'temperatureMax', 'temperatureMaxTime',
       'apparentTemperatureMin', 'apparentTemperatureMinTime',
       'apparentTemperat

Diatas adalah list dari kolom yang ada didalam dataset

## Exploratory Data Analysis

Memisahkan data yang terdapat nilai null dan tidak null

In [41]:
data_train = uber_lyft_df[uber_lyft_df['price'].notnull()]
data_test = uber_lyft_df[uber_lyft_df['price'].isnull()]
print(data_train.shape)
print(data_test.shape)

(637976, 57)
(55095, 57)


## Data Preprocessing

melakukan train test split pada data train

In [28]:
x_train, x_test, y_train, y_test = train_test_split(data_train[num_columns], data_train['price'], test_size=0.2, random_state=1)

## Model Definition

## Model Evaluation

## Model Inference

## Pengambilan Kesimpulan

## Referensi

Hidayat, A., 2012, Pengertian Analisis Regresi Korelasi Dan Cara Hitung, Statistikian, Retrieved from https://www.statistikian.com/2012/08/analisis-regresi-korelasi.html